core/auditlog/auditlog.go

```diff
@@ -0,0 +1,240 @@
+type ResponseArgs struct {
+	RequestID uint64
+	Errors    []*Error
+}
+
+// Error holds the details of an error sent back from the API.
+type Error struct {
+	Message string `json:"message"`
+	Code    string `json:"code"`
+}
+
+// Record is the top-level entry type in an audit log, which serves as
+// a type discriminator. Only one of Call/Request/Response should be set.
+type Record struct {
+	Call     *Call           `json:"call,omitempty"`
+	Request  *FacadeRequest  `json:"request,omitempty"`
+	Response *FacadeResponse `json:"response,omitempty"`
+}
+
+// AuditLog represents something that can store calls, requests and
+// responses somewhere.
+type AuditLog interface {
+	AddCall(c Call) error
+	AddRequest(r FacadeRequest) error
+	AddResponse(r FacadeResponse) error
+	Close() error
+}
+
+// Recorder records method calls for a specific API connection.
+type Recorder struct {
+	log          AuditLog
+	connectionID string
+	callID       string
+}
+
+// NewRecorder creates a Recorder for the connection described (and
+// stores details of the initial call in the log).
+func NewRecorder(log AuditLog, c CallArgs) (*Recorder, error) {
+	callID := newCallID()
+	connectionID := idString(c.ConnectionID)
+	err := log.AddCall(Call{
+		CallID:       callID,
+		ConnectionID: connectionID,
+		Who:          c.Who,
+		What:         c.What,
+		When:         c.When.Format(time.RFC3339),
+		ModelName:    c.ModelName,
+		ModelUUID:    c.ModelUUID,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &Recorder{
+		log:          log,
+		callID:       callID,
+		connectionID: connectionID,
+	}, nil
+}
+
+// AddRequest records a method call to the API.
+func (r *Recorder) AddRequest(m RequestArgs) error {
+	return errors.Trace(r.log.AddRequest(FacadeRequest{
+		CallID:       r.callID,
+		ConnectionID: r.connectionID,
+		RequestID:    m.RequestID,
+		Facade:       m.Facade,
+		Method:       m.Method,
+		Version:      m.Version,
+		Args:         m.Args,
+	}))
+}
+
+// AddResponse records the result of a method call to the API.
+func (r *Recorder) AddResponse(m ResponseArgs) error {
+	return errors.Trace(r.log.AddResponse(FacadeResponse{
+		CallID:       r.callID,
+		ConnectionID: r.connectionID,
+		RequestID:    m.RequestID,
+		Errors:       m.Errors,
+	}))
+}
+
+func newCallID() string {
+	buf := make([]byte, 8)
+	rand.Read(buf) // Can't fail
+	return hex.EncodeToString(buf)
+}
+
+type AuditLogFile struct {
+	fileLogger io.WriteCloser
+}
+
+// NewLogFile returns an audit entry sink which writes to an audit.log
+// file in the specified directory.
+func NewLogFile(logDir string) *AuditLogFile {
+	logPath := filepath.Join(logDir, "audit.log")
+	if err := primeLogFile(logPath); err != nil {
+		// This isn't a fatal error so log and continue if priming
+		// fails.
+		logger.Errorf("Unable to prime %s (proceeding anyway): %v", logPath, err)
+	}
+
+	return &AuditLogFile{
+		fileLogger: &lumberjack.Logger{
+			Filename:   logPath,
+			MaxSize:    300, // MB
```

Comment:
todo - make these configurable?

Reply:
I'll make them parameters of NewLogFile.